osutil/epoll/epoll.go

```diff
@@ -0,0 +1,201 @@
+// Package epoll contains a thin wrapper around the epoll(7) facility.
+//
+// Using epoll from Go is unusual as the language provides facilities that
+// normally make using it directly pointless. Epoll is strictly required for
+// unusual kernel interfaces that use event notification but don't implement
+// file descriptors that provide usual read/write semantics.
+
+package epoll
+
+import (
+	"fmt"
+	"runtime"
+	"strings"
+	"sync/atomic"
+	"time"
+
+	"golang.org/x/sys/unix"
+)
+
+// Readiness is the bit mask of aspects of readiness to monitor with epoll.
+type Readiness int
+
+const (
+	// Readable indicates readiness for reading (EPOLLIN).
+	Readable Readiness = unix.EPOLLIN
+	// Writable indicates readiness for writing (EPOLLOUT).
+	Writable Readiness = unix.EPOLLOUT
+)
+
+// String returns readable representation of the readiness flags.
+func (r Readiness) String() string {
+	frags := make([]string, 0, 2)
+	if r&Readable != 0 {
+		frags = append(frags, "Readable")
+	}
+	if r&Writable != 0 {
+		frags = append(frags, "Writable")
+	}
+	return strings.Join(frags, "|")
+}
+
+// Epoll wraps a file descriptor which can be used for I/O readiness notification.
+type Epoll struct {
+	fd                int
+	registeredFdCount int32 // read/modify using helper functions
+}
+
+// Open opens an event monitoring descriptor.
+func Open() (*Epoll, error) {
+	fd, err := unix.EpollCreate1(unix.EPOLL_CLOEXEC)
+	if err != nil {
+		return nil, fmt.Errorf("cannot open epoll file descriptor: %w", err)
+	}
+	e := &Epoll{
+		fd:                fd,
+		registeredFdCount: 0,
+	}
+	runtime.SetFinalizer(e, func(e *Epoll) {
+		if e.fd != -1 {
+			e.Close()
+		}
+	})
+	return e, nil
+}
+
+// Close closes the event monitoring descriptor.
+func (e *Epoll) Close() error {
+	runtime.SetFinalizer(e, nil)
+	fd := e.fd
+	e.fd = -1
+	e.zeroRegisteredFdCount()
+	return unix.Close(fd)
+}
+
+// Fd returns the integer unix file descriptor referencing the open file.
+func (e *Epoll) Fd() int {
+	return e.fd
+}
+
+// RegisteredFdCount returns the number of file descriptors which are currently
+// registered to the epoll instance.
+func (e *Epoll) RegisteredFdCount() int {
+	return int(atomic.LoadInt32(&e.registeredFdCount))
+}
+
+func (e *Epoll) incrementRegisteredFdCount() {
+	atomic.AddInt32(&e.registeredFdCount, 1)
+}
+
+func (e *Epoll) decrementRegisteredFdCount() {
+	atomic.AddInt32(&e.registeredFdCount, -1)
+}
+
+func (e *Epoll) zeroRegisteredFdCount() {
+	atomic.StoreInt32(&e.registeredFdCount, 0)
+}
+
+// Register registers a file descriptor and allows observing speicifc I/O readiness events.
+//
+// Please refer to epoll_ctl(2) and EPOLL_CTL_ADD for details.
+func (e *Epoll) Register(fd int, mask Readiness) error {
+	err := unix.EpollCtl(e.fd, unix.EPOLL_CTL_ADD, fd, &unix.EpollEvent{
+		Events: uint32(mask),
+		Fd:     int32(fd),
+	})
+	if err != nil {
+		return err
+	}
+	e.incrementRegisteredFdCount()
```

Comment:
nitpick: in principle this should go before the EpollCtl and then have a deregister in the error case